src/coreclr/jit/morph.cpp

```diff
@@ -13740,6 +13732,46 @@ GenTree* Compiler::fgOptimizeAddition(GenTreeOp* add)
         return op1;
     }
 
+    // Reduce local addresses: ADD(ADDR(LCL_VAR), OFFSET) => ADDR(LCL_FLD OFFSET).
+    // TODO-ADDR: do ADD(LCL_FLD/VAR_ADDR, OFFSET) => LCL_FLD_ADDR instead.
+    //
+    if (opts.OptimizationEnabled() && fgGlobalMorph && op1->OperIs(GT_ADDR) && op2->IsCnsIntOrI() &&
+        op1->AsUnOp()->gtGetOp1()->OperIs(GT_LCL_VAR, GT_LCL_FLD))
+    {
+        GenTreeUnOp*         addrNode   = op1->AsUnOp();
+        GenTreeLclVarCommon* lclNode    = addrNode->gtGetOp1()->AsLclVarCommon();
+        GenTreeIntCon*       offsetNode = op2->AsIntCon();
+        if (FitsIn<uint16_t>(offsetNode->IconValue()))
+        {
+            unsigned offset = lclNode->GetLclOffs() + static_cast<uint16_t>(offsetNode->IconValue());
+
+            if (FitsIn<uint16_t>(offset) && (offset < lvaLclExactSize(lclNode->GetLclNum())))
+            {
+                // Compose the field sequence: [LCL, ADDR, OFFSET].
+                FieldSeqNode* fieldSeq           = lclNode->GetFieldSeq();
+                FieldSeqNode* zeroOffsetFieldSeq = nullptr;
+                if (GetZeroOffsetFieldMap()->Lookup(addrNode, &zeroOffsetFieldSeq))
+                {
+                    fieldSeq = GetFieldSeqStore()->Append(fieldSeq, zeroOffsetFieldSeq);
+                    GetZeroOffsetFieldMap()->Remove(addrNode);
+                }
+                fieldSeq = GetFieldSeqStore()->Append(fieldSeq, offsetNode->gtFieldSeq);
+
+                // Types of location nodes under ADDRs do not matter. We arbitrarily choose TYP_UBYTE.
+                lclNode->ChangeType(TYP_UBYTE);
```

Comment:
Maybe have it `TYP_UNDEF` so we would catch issues if it is accidentally used in future?

Reply:
I considered it, the first version of the change used `TYP_VOID` here (one small VN tweak was needed but otherwise it worked ok).

Ultimately my judgement was that we're at a lower risk of surprises if we don't introduce "new" IR shapes and so I backed that out.